Travis -> GitHub workflows

GeorgianaElena opened the pull request.

consideRatio commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,41 @@
+  push:
+    # FIXME: use the release event instead?
+    tags: [0-9]+.[0-9]+.[0-9]+
+    branches: master
```
This will not support any beta releases etc though, its worth making a note about or add support for, for example by appedning .* in the end.

consideRatio commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,41 @@
+
+jobs:
+  build-n-publish:
+    name: Build and publish on tags to PyPI
```
It will always build, but only also do the publish step on (specific) tags.

@minrk did the same in https://github.com/jupyterhub/oauthenticator/pull/383. If possible I'd like to get the "dummy publish" to run as a test under a workflow not named Release/Publish or similarly.

I think it will avoid reactions like "Oh did my push to master trigger a release!?!", to then be followed by mandatory precautionary inspection leading to "oh, the specific step to push wasn't triggered".

Also, if someone then push a tag that didn't trigger the workflow (1.0.0b1 for example), but it triggered by a normal push of a commit to the main branch, then that user would see a release/publish workflow be triggered and succeeding (but as master), so that user would in this case think it was all alright, but then not see any PyPI package because it never ran on the tagged commit, only on the master.

And perhaps finally, a user that wants to make use of a development release or similarly may inspect and see a Publish workflow and assume there is a development release on pushes to master or similarly.

## Summary
I consider the value of always testing the build of the package, and personally think it adds confusion to do so but only conditionally publish to PyPi under something named Publish or Release.

consideRatio commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,41 @@
+        TWINE_PASSWORD: ${{ secrets.PYPI_APIKEY }}
+      run: |
+        python setup.py sdist bdist_wheel
+        twine upload dist/*
```
This step currently will always install pip, setuptools, wheel, twine, but only if we push a tag will we build and publish to PyPI.

We should always let the package step run

consideRatio commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,56 @@
+          pip freeze
+      - name: Run tests
+        run: |
+          pytest -v tests/${{ matrix.tests }} --color=yes --cov=jupyterhub_traefik_proxy
```
This makes us run 1 job per test file and python version, why not run all test files here and make it run 1 job per python version?

GeorgianaElena replied:
I was experimenting with running them in parallel because I noticed that the proxy and the installer tests take a bit of time. Do you think it looks "hacky" and I shouldn't be doing it?
An example of the jobs that are run:
https://github.com/GeorgianaElena/traefik-proxy/actions/runs/357984398

consideRatio replied:
By creating additional jobs, we create additional overhead. So, by splitting it in four, we do the installation of dependencies four times instead of one. In time that adds ~100s or so of total compute time for the entire workflow. We gained in the total time to complete the workflow because we used separate jobs running in parallel without queuing up.

For me, the following things matter:
1. The total time to complete a workflow
2. The ease of understanding how things work quickly and reduce cognitive work for developers
3. The time to run quick test for quick feedback. For example, its good to run the basic linting test before you start the long and slow tests.
4. The total resources used by our CI system
5. The reliability of the outcome assuming things change, where for example github actions reduce our parallel runners to just maximum or similarly.

- If the parallel workers available drop to ~4, then this kind of optimization is wasted.
- The biggest benefit of total time to complete a workflow comes from splitting the test runs into two equally large, so a split of 5:5 rather than 1:1:1:7 for example.

Since the proxy file's tests add up to ~14 minutes and the other tests add up to less than a minute, I suggest to not do this but instead select the tests that are slow with a pytest marker, and then you create two separate pytests jobs. One running `pytest -m slow` and one running `pytest -m "not slow"`. The outcome of this is that you ensure that the fast tests will run and complete in ~1 min, and then for the full test suite to run, you always need to wait a long time.

If you want to optimize under the assumption that tests will keep being able to run in massive parallelism, then you could split apart the tests into X groups of equal size and use that unlimited parallelism to your advantage.

I think it is acceptable as it is though, but I would ask for a comment explaining that the motivation for this is to parallelize some tests and ensure that the quick tests complete quickly and not get stuck behind slow tests.

consideRatio replied:
If you mark certain tests as slow, for example all tests in the proxy folder or just some in general, you can sort them with this hook you can defined in conftest.py: https://docs.pytest.org/en/latest/reference.html#pytest.hookspec.pytest_collection_modifyitems, for example based on the markers you have.

GeorgianaElena replied:
Thank you for sharing these awesome ideas, Erik. I wasn't entirely sure what I was doing or trying to achieve, other than playing around with gha, jobs and try to make things faster :sweat_smile:

I'm now ordering the tests as per your suggestion, leaving the slow ones at the end, which I think makes a lot of sense to have. Running them in parallel as I did (per test file) wasn't helping much, since the proxy tests are the ones that are slow and numerous, so I was only saving a tini tiny bit of time from running the faster ones in parallel with this.

The best thing to do would probably be splitting the proxy tests somehow, but I'll leave that for another PR.
I believe having them ordered at this point makes more sense. 

TY :heart:

consideRatio commented:
@GeorgianaElena I've added a pypi_password in this repo as a github repo secret you can reference like @minrk has done in jupyterhub/oauthenticator

consideRatio commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,41 @@
+# but only publish to PyPI on tags
+on:
+  push:
+  pull_request:
```
https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet

I think the tags condition were interesting, but I think you got the regex syntax wrong in a recent commit.

Here is what I think would work and would be generally functional for all our repos in the jupyterhub org.

```
v?[0-9]+.[0-9]+.[0-9]+**
```

GeorgianaElena replied:
I removed regex because gha was complaining about syntax :( Will try more to make it work if you too think it's interesting to have it (I liked it too when I saw it in a blog post :heart: )

> v?[0-9]+.[0-9]+.[0-9]+**

`v?`: nice  :+1:
`**`: according to the [cheat-sheet](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet) (thanks for sharing it :heart:), one `*` won't match the `/` character, and `**` will match that. Do we want to match `/`? Is this commonly used in version names?

consideRatio replied:
I became uncertain and confused and thought that a single `*` would represent a single character only, but it turns out that it was matching many. I got mislead reading "The * wildcard matches any character, but does not match slash (/)." to think it was just a single character being matched.

So, then the question of `*` vs `**` is just a matter of inclusion of `/` or not, and I don't think we will ever experience a difference between these so choose either!

consideRatio commented on tests/test_proxy.py:
```diff
@@ -4,8 +4,8 @@
-# Mark all tests in this file as asyncio
-pytestmark = pytest.mark.asyncio
+# Mark all tests in this file as asyncio and slow
+pytestmark = [pytest.mark.asyncio, pytest.mark.slow]
```
Ah, new things I learned on how to mark tests, nice!

consideRatio commented on tests/conftest.py:
```diff
@@ -7,12 +7,29 @@
+    if config.getoption('--slow-last'):
+        items.sort(key=by_slow_marker)
+
+
```
Oooh nice solution! :heart_eyes:

consideRatio reviewed (approved):
Excellent work on this, thanks for your patience with my review comments @GeorgianaElena!

GeorgianaElena commented:
Thanks for taking the time to review this @consideRatio :heart: Your reviews are always an opportunity to learn about a lot.

I believe there's still a couple of things left to fix:
- if we add the tags condition, the build won't run on every push. It would be nice to be able to keep building on every push and only publish on tags with that match our regex. I'm thinking we have two options to do this:
    * separate build and push in two yml files
    * see if the condition `startsWith` of `pypa/gh-action-pypi-publish@v1.4.1`accepts regex (I didn't find docs about it, but I believe it's worth trying anyway).
- the build in the release job actually fails right now (see the run from yesterday: https://github.com/GeorgianaElena/traefik-proxy/runs/1391443808?check_suite_focus=true). It might have something to do with having `setuptools` listed in the `pyproject.toml` :woman_shrugging: (though after reading https://snarky.ca/what-the-heck-is-pyproject-toml/ it doesn't have a lot of sense)

What do you think?

consideRatio commented:
> if we add the tags condition, the build won't run on every push.

Do you mean that if we add the tags condition, we won't run tests on pushes of tags not having a version constraint? I think that is fine.

I suggest to have the build package step as part of a test in the test suite, duplicated from the publishing workflow. If it is part of the publishing workflow that always run, I know that for me, it force me to inspect "did it REALLY publish, or did it just run the build package steps?" whenever I really want to know if it actually published during release.

My wishes in order of importance
1. To only trigger the publish workflow if we actually publish
2. To test the python package packaging
3. To not break DRY with regards to testing the python package packaging
4. To test the python package packaging on non-version like tags.

If 2-4 would add notable complexity or similar, I'd personally prefer to not bother about them. I have only felt the need to test the packaging part when I develop the CI for the packaging part, and even then, a test would not capture the validity of environment variables for deployment keys to PyPI etc.

consideRatio commented:
## Build error

I think it is caused doing `python -m pip install --upgrade setuptools pip` and then `python -m build --sdist --wheel .`, and it makes setuptools install twice or similar because of the `pyproject.toml` contain setuptools as a build-system dependency. I suggest to try removing the explicit installation of setuptools in a `pip install` step.

```yaml
    - name: Install build package
      run: |
        python -m pip install --upgrade setuptools pip
        pip install build
        pip freeze
    - name: Build release
      run: |
        python -m build --sdist --wheel .
        ls -l dist
```

## README badge
The README contain a badge pointing to travis still. Here is a badge in markdown for you to use, note it reference a specific workflow, and it contains the case-sensitive URL encoded name of the workflow which is currently `Run tests`.


```
[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/jupyterhub/traefik-proxy/Run%20tests?logo=github)](https://github.com/jupyterhub/traefik-proxy/actions)
```

GeorgianaElena commented:
Thanks for the feedback @consideRatio. I added a few commits in which:
- Added the new badge to readme :tada: 
- Added the pkg build step to the python 3.8 test suite
- Removed completely the build requirements (`wheel` and `setuptools`) from `pyproject.toml` as they get installed as a `build` package requirement and otherwise it complains with:
    ```
    ERROR: Double requirement given: setuptools (from -r /tmp/build-reqs-y789h1lu.txt (line 3)) (already in setuptools>=40.8.0 (from -r /tmp/build-reqs-y789h1lu.txt (line 1)), name='setuptools')
    ```

New observation:
One proxy test (the first one that starts traefik) is flaky as it sometimes fails to wait for traefik to load it's config. I increased the start timeout for that proxy but still didn't manage to make it stable, so suggest will leave as it is (at least for this PR) and try different approaches later on to fix this.

What do you think @consideRatio ? Is there anything else I could do to improve this PR?

consideRatio commented:
>  Removed completely the build requirements (wheel and setuptools) from pyproject.toml as they get installed as a build package requirement and otherwise it complains with:

I'm not knowledgable about pyproject.toml, but I think it could make more sense to remove the dedicated installation step of `setuptools` from the `pip install --upgrade setuptools pip` statement in the CI system, and let the pyproject.toml help it get installed. But, I really know to little about this to feel confident, but there is a warning bell saying that we may have made something that made sense now only work in the CI system but not work as intended outside the CI system.

> One proxy test (the first one that starts traefik) is flaky as it sometimes fails to wait for traefik to load it's config. I increased the start timeout for that proxy but still didn't manage to make it stable, so suggest will leave as it is (at least for this PR) and try different approaches later on to fix this.

Sounds good!

> What do you think @consideRatio ? Is there anything else I could do to improve this PR?

No =) Give the pyproject.toml part a think and then merge when have considered it :D

manics commented:
I spent a few hours fighting with `pyproject.toml` over the weekend when I was updating jupyter-offlinenotebook.

It's related to [PEP 517](https://www.python.org/dev/peps/pep-0517/) and/or [PEP 518](https://www.python.org/dev/peps/pep-0518/).

From my limited understanding whether or not you need the build dependencies in `pyproject.toml` depends on how you build and install this package, since the build may be done in an _isolated_ environment. `pip install` may behave differently from `python setup.py`, so we need to decide how people are likely to use this repo, and/or what we want to test or support. I concluded in my case it was necessary to have the build-dependencies in `pyproject.toml`.

GeorgianaElena commented:
Thanks a lot @manics for pointing me in the right direction :heart:

The duplicate error was from the `build` package, which was reinserting the `setuptools` and `wheel` to the list of the requirements when no [`build-backend`](https://www.python.org/dev/peps/pep-0517/#source-trees) was provided in `pyproject.toml` (even though `requires` was there). This is fixed now, but not yet released, so I decided to explicitly specify it in the toml file.

Thanks a lot for your help @consideRatio and @manics. I'm going to merge this now.
